src/Polysemy/Internal/TH/Effect.hs

```diff
@@ -75,135 +79,185 @@ makeSem = genFreer True
+      conArgs           = applySubstitution eqPairs constructorFields
+
+  (effTArgs, [mArg, resArg]) <- case splitFromEnd 2 allEffTArgs of
+      r@(_, [_, _]) -> return r
+      _             -> fail $ show $ missingEffTArgs typeName
+  mName <-
+    -- May happen when used on data families
+    maybe (fail $ show $ mArgNotVar typeName mArg) return $ tvName' mArg
+
+  return CLInfo {
+    typeName, funName, conName, effTArgs, mName, resArg, conCxt, conArgs
+    }
+  where
+    mArgNotVar name mArg
+      =   text "Monad argument ‘" <> ppr mArg <> text "’ in effect ‘"
+          <> ppr name <> text "’ is not a type variable"
+
+    missingEffTArgs name
+      =   text "Effect ‘" <> ppr name
+          <> text "’ has not enough type arguments"
+      $+$ nest 4
+          (   text "At least monad and result argument are required, e.g.:"
+          $+$ nest 4
+              (   text ""
+              $+$ ppr (DataD [] base args Nothing [] []) <+> text "..."
+              $+$ text ""
+              )
+          )
+      where
+        base = mapNameBase id name
+        args = PlainTV . mkName <$> ["m", "a"]
+
+------------------------------------------------------------------------------
+-- | Generates signature for lifting function and type arguments to apply in
+-- it's body on effect's data constructor.
+genSig :: ConLiftInfo -> Q Dec
+genSig CLInfo {
+    typeName, funName, effTArgs, mName, resArg, conCxt, conArgs
+  } = do
+  r <- newName "r"
+  let memberCxt  = ConT ''Member `AppT` eff `AppT` VarT r
+      eff        = foldl' AppT (ConT typeName) effTArgs
+      -- TODO: can use of 'm' blow up somehow?
+      funArgs    = replaceMArg mName r conArgs
+      returnType = ConT ''Sem `AppT` VarT r `AppT` resArg
+
+  return $ SigD funName $ quantifyType $
+    ForallT [] (memberCxt : conCxt) $ foldArrows $ funArgs ++ [returnType]
+
+  where
+    -- Replaces use of @m@ in type with @Sem r@.
+    replaceMArg m r = applySubstitution
+                    $ M.singleton m (ConT ''Sem `AppT` VarT r)
+
+------------------------------------------------------------------------------
 -- | Builds a function definition of the form @x a b c = send $ X a b c@.
-genDecl :: Con -> Q [Dec]
-genDecl (ForallC _       _     con) = genDecl con
-genDecl (GadtC   [cName] tArgs _  ) = do
-  let fnName = getDeclName cName
-  let arity  = length tArgs - 1
-  dTypeVars <- forM [0 .. arity] $ const $ newName "a"
-  pure $
-    [PragmaD (InlineP fnName Inlinable ConLike AllPhases)
-    , FunD fnName . pure $ Clause
-        (VarP <$> dTypeVars)
-        (NormalB . AppE (VarE 'send) $ foldl
-          (\b -> AppE b . VarE)
-          (ConE cName)
-          dTypeVars
-        )
-        []
+genDec :: ConLiftInfo -> Q [Dec]
+-- TODO: apply effect's type arguments to avoid "bad ambiguity"
+genDec CLInfo { funName, conName, conArgs } = do
+  funArgsNames <- replicateM (length conArgs) $ newName "x"
+
+  return [
```

Comment:
Nit: `[`s on their own line, indented two; align with `,`s, eg:

```
foo = 
  [ bar
  , baz
  ]
```

Here and elsewhere

Reply:
Ok